From 8f5871a3030d2615b157d6a4748a18c7b104a221 Mon Sep 17 00:00:00 2001
From: Albin B <albin.bala.krishnan@intel.com>
Date: Wed, 17 Dec 2014 13:56:07 +0530
Subject: [PATCH 1/1] pmic-ccsm: Fix for OTG throttling/de-throttling issue

As the charger driver is not handling notifier events for vbus enable/disable
and using usb_phy->set_vbus function to enable/disable the vbus, added set_vbus
function to enable/disable the otg when the cur_state of the cooling device is
getting changed.

Change-Id: I411408dbceae47292c96c59774f92af8cecbc470
Tracked-on: https://jira01.devtools.intel.com/browse/IMINAN-14069
Signed-off-by: Albin B <albin.bala.krishnan@intel.com>
---
 drivers/power/intel_pmic_ccsm.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/power/intel_pmic_ccsm.c b/drivers/power/intel_pmic_ccsm.c
index a490da9..b4aafc9 100644
--- a/drivers/power/intel_pmic_ccsm.c
+++ b/drivers/power/intel_pmic_ccsm.c
@@ -998,6 +998,8 @@ static void handle_pwrsrc_interrupt(u16 int_reg, u16 stat_reg)
 	if ((int_reg & BIT_POS(PMIC_INT_USBIDDET)) &&
 			(chc.vbus_state == VBUS_ENABLE)) {
 		mask = !!(stat_reg & BIT_POS(PMIC_INT_USBIDDET));
+		if (chc.otg->set_vbus)
+			chc.otg->set_vbus(chc.otg, true);
 		atomic_notifier_call_chain(&chc.otg->notifier,
 				USB_EVENT_ID, &mask);
 	}
@@ -1524,14 +1526,19 @@ static int vbus_set_cur_state(struct thermal_cooling_device *tcd,
 	 */
 	mutex_lock(&pmic_lock);
 	if ((pmic_get_usbid() == RID_GND) && (chc.vbus_state != new_state)) {
-		if (!new_state)
+		if (!new_state) {
+			if (chc.otg->set_vbus)
+				chc.otg->set_vbus(chc.otg, true);
 			atomic_notifier_call_chain(&chc.otg->notifier,
 						USB_EVENT_ID,
 						NULL);
-		else
+		} else {
+			if (chc.otg->set_vbus)
+				chc.otg->set_vbus(chc.otg, false);
 			atomic_notifier_call_chain(&chc.otg->notifier,
 						USB_EVENT_NONE,
 						NULL);
+		}
 	}
 
 	chc.vbus_state = new_state;
-- 
1.9.1

